spanner/client.go

```diff
@@ -121,6 +124,131 @@ func (c *Client) ClientID() string {
 	return c.sc.id
 }
 
+func createGCPMultiEndpoint(cfg *grpcgcp.GCPMultiEndpointOptions, config ClientConfig, opts ...option.ClientOption) (*grpcgcp.GCPMultiEndpoint, error) {
+	if cfg.GRPCgcpConfig == nil {
+		cfg.GRPCgcpConfig = &grpcgcppb.ApiConfig{}
+	}
+	if cfg.GRPCgcpConfig.Method == nil || len(cfg.GRPCgcpConfig.Method) == 0 {
+		cfg.GRPCgcpConfig.Method = []*grpcgcppb.MethodConfig{
+			{
+				Name: []string{"/google.spanner.v1.Spanner/CreateSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BIND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/BatchCreateSessions"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BIND,
+					AffinityKey: "session.name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/DeleteSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_UNBIND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/GetSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BOUND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{
+					"/google.spanner.v1.Spanner/BeginTransaction",
+					"/google.spanner.v1.Spanner/Commit",
+					"/google.spanner.v1.Spanner/ExecuteBatchDml",
+					"/google.spanner.v1.Spanner/ExecuteSql",
+					"/google.spanner.v1.Spanner/ExecuteStreamingSql",
+					"/google.spanner.v1.Spanner/PartitionQuery",
+					"/google.spanner.v1.Spanner/PartitionRead",
+					"/google.spanner.v1.Spanner/Read",
+					"/google.spanner.v1.Spanner/Rollback",
+					"/google.spanner.v1.Spanner/StreamingRead",
+				},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BOUND,
+					AffinityKey: "session",
+				},
+			},
+		}
+	}
+	// Append emulator options if SPANNER_EMULATOR_HOST has been set.
+	if emulatorAddr := os.Getenv("SPANNER_EMULATOR_HOST"); emulatorAddr != "" {
+		emulatorOpts := []option.ClientOption{
+			option.WithEndpoint(emulatorAddr),
+			option.WithGRPCDialOption(grpc.WithTransportCredentials(insecure.NewCredentials())),
+			option.WithoutAuthentication(),
+			internaloption.SkipDialSettingsValidation(),
+		}
+		opts = append(opts, emulatorOpts...)
```

Comment:
Note that mutating a var args parameter (opts) is dangerous.  This code should be making a copy first.

https://go.dev/play/p/UAEsaz4xxXg

This is a contrived example to illustrate what's happening, but there is real risk in mutating a variadic argument, especially for exported APIs.

Reply:
Also fixing in #11151